### PR TITLE
PIM-5800: Improve how to add a new step cookbook

### PR DESCRIPTION
**Pull request by solivier**

### CODE REVIEW

File: src/Acme/Bundle/NotifyConnectorBundle/JobParameters/ConstraintCollectionProvider/ProductCsvExportNotify.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+namespace Acme\Bundle\NotifyConnectorBundle\JobParameters\ConstraintCollectionProvider;
+
+use Akeneo\Component\Batch\Job\JobInterface;
+use Akeneo\Component\Batch\Job\JobParameters\ConstraintCollectionProviderInterface;
+use Pim\Component\Connector\Validator\Constraints\Channel;
+use Pim\Component\Connector\Validator\Constraints\FilterData;
+use Pim\Component\Connector\Validator\Constraints\FilterStructureAttribute;
+use Pim\Component\Connector\Validator\Constraints\FilterStructureLocale;
+use Symfony\Component\Validator\Constraints\Collection;
+use Symfony\Component\Validator\Constraints\NotBlank;
```
Comment (skeleton) [inline src/Acme/Bundle/NotifyConnectorBundle/JobParameters/ConstraintCollectionProvider/ProductCsvExportNotify.php]: remove useless namespace please

File: src/Acme/Bundle/NotifyConnectorBundle/JobParameters/FormConfigurationProvider/ProductCsvExportNotify.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+     * {@inheritdoc}
+     */
+    public function getFormConfiguration()
+    {
+        $csvFormOptions = array_merge($this->productCsvExport->getFormConfiguration(), [
+            'url' => [
+                'type' => 'url',
+                'options' => [
+                    'required' => true,
+                    'label'    => 'test',
```
Comment (skeleton) [inline src/Acme/Bundle/NotifyConnectorBundle/JobParameters/FormConfigurationProvider/ProductCsvExportNotify.php]: test ?

File: cookbook/import_export/create-custom-step.rst [hunk 3]
```diff
@@ -15,36 +15,88 @@ We will begin by creating a NotifyStep with its configuration and a doExecute me
...
-----------------------------------
+And don't forget to declare this class as a service, with the proper job name as parameter:
 
-Add your step element to ``steps.yml`` to ensure that it is loaded and processed by ``DependencyInjection/AcmeNotifyConnectorExtension``:
+.. literalinclude:: ../../src/Acme/Bundle/NotifyConnectorBundle/Resources/config/job_parameters.yml
+   :language: yaml
+   :linenos:
 
-.. literalinclude:: ../../src/Acme/Bundle/NotifyConnectorBundle/Resources/config/steps.yml
+Configure the job form constraint
```
Comment (nidup) [inline cookbook/import_export/create-custom-step.rst]: "Configure the job form constraint" -> "Configure the job constraint" not only for the form, used during execution etc

File: cookbook/import_export/create-custom-step.rst [hunk 4]
```diff
@@ -15,36 +15,88 @@ We will begin by creating a NotifyStep with its configuration and a doExecute me
...
+
+And don't forget to declare this class as a service, with the proper job name as parameter:
+
+.. literalinclude:: ../../src/Acme/Bundle/NotifyConnectorBundle/Resources/config/job_constraints.yml
    :language: yaml
    :linenos:
 
-Configure our new Job
----------------------
+Configure the job form default values
```
Comment (nidup) [inline cookbook/import_export/create-custom-step.rst]: idem than previous, i would start by default values, then constraint then ui form

File: cookbook/ui.rst [hunk 5]
```diff
@@ -5,7 +5,7 @@ This chapter gives details on how to customize existing UI to fit your customer'
 
 .. toctree::
 
-   ui_customization/add_a_new_tab_to_an_entity_edit_form
+   ui_customization/add_a_neçaw_tab_to_an_entity_edit_form
```
Comment (willy-ahva) [inline cookbook/ui.rst]: fixçawe neçaw_tab plsçaw

Comment (skeleton) [inline cookbook/ui.rst]: typo :)

File: src/Acme/Bundle/NotifyConnectorBundle/JobParameters/ConstraintCollectionProvider/ProductCsvExportNotify.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+use Symfony\Component\Validator\Constraints\Collection;
+use Symfony\Component\Validator\Constraints\Url;
+
+class ProductCsvExportNotify implements ConstraintCollectionProviderInterface
+{
+    /** @var ConstraintCollectionProviderInterface */
+    protected $productCsvExport;
+
+    /** @var array */
+    protected $supportedJobNames;
```
Comment (jjanvier) [inline src/Acme/Bundle/NotifyConnectorBundle/JobParameters/ConstraintCollectionProvider/ProductCsvExportNotify.php]: I wouldn't do that
here we are in a project
keep things simple

File: src/Acme/Bundle/NotifyConnectorBundle/JobParameters/ConstraintCollectionProvider/ProductCsvExportNotify.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+        return new Collection(['fields' => $constraintFields]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supports(JobInterface $job)
+    {
+        return in_array($job->getName(), $this->supportedJobNames);
```
Comment (jjanvier) [inline src/Acme/Bundle/NotifyConnectorBundle/JobParameters/ConstraintCollectionProvider/ProductCsvExportNotify.php]: return $job->getName() === 'my_job_name' 

imho

Comment (jjanvier) [inline src/Acme/Bundle/NotifyConnectorBundle/JobParameters/ConstraintCollectionProvider/ProductCsvExportNotify.php]: same for all the others classes

**Comment by jjanvier** [comment]
to ease integration, I would do a single class that implements all the 3 important interfaces (default values, constraints validation, ui form conf)

File: cookbook/import_export/create-custom-step.rst [hunk 8]
```diff
@@ -15,36 +15,88 @@ We will begin by creating a NotifyStep with its configuration and a doExecute me
 .. literalinclude:: ../../src/Acme/Bundle/NotifyConnectorBundle/Step/NotifyStep.php
    :language: php
    :linenos:
-   :lines: 1-4,10-
 
-Create our Step Element
------------------------
+Define the Step as Service
```
Comment (jjanvier) [inline cookbook/import_export/create-custom-step.rst]: as A service?

**Comment by nidup** [comment]
@jjanvier ok for a single class for providers, and a simpler support method :)

File: cookbook/import_export/create-custom-step.rst [hunk 9]
```diff
@@ -15,36 +15,92 @@ We will begin by creating a NotifyStep with its configuration and a doExecute me
...
+Add your step element to ``steps.yml`` to ensure that it is loaded and processed by ``DependencyInjection/AcmeNotifyConnectorExtension``:
+
+.. literalinclude:: ../../src/Acme/Bundle/NotifyConnectorBundle/Resources/config/steps.yml
+   :language: yaml
+   :linenos:
+
+Configure our new Job
+---------------------
+
+In ``Resources/config/jobs.yml``, we use a first step to export products in csv and we configure the second one to send a notification:
```
Comment (rybus) [inline cookbook/import_export/create-custom-step.rst]: CSV

File: cookbook/import_export/create-custom-step.rst [hunk 10]
```diff
@@ -15,36 +15,92 @@ We will begin by creating a NotifyStep with its configuration and a doExecute me
...
+.. literalinclude:: ../../src/Acme/Bundle/NotifyConnectorBundle/Resources/config/jobs.yml
+   :language: yaml
+   :linenos:
+
+Configure the job parameters
+----------------------------
+
+In order to setup the form we use a form configuration provider.
+
+First of all you need to create a new class that will contains the job parameters. This class should implement the following interfaces:
```
Comment (rybus) [inline cookbook/import_export/create-custom-step.rst]: that will contain (no 's')

File: cookbook/import_export/create-custom-step.rst [hunk 11]
```diff
@@ -15,36 +15,92 @@ We will begin by creating a NotifyStep with its configuration and a doExecute me
...
+   :lines: 74-79
+
+And don't forget the support method with your job name ``csv_product_export_notify``.
+
+.. literalinclude:: ../../src/Acme/Bundle/NotifyConnectorBundle/JobParameters/ProductCsvExportNotify.php
+   :language: php
+   :linenos:
+   :lines: 85-88
+
+And don't forget to declare this class as a service, with the proper job name as parameter:
```
Comment (rybus) [inline cookbook/import_export/create-custom-step.rst]: "And don't forget" is used twice, maybe just "Add the support method with your [...]" and for the second one "Declare this class as a service" to keep it simple ?

File: cookbook/import_export/create-custom-step.rst [hunk 12]
```diff
@@ -15,36 +15,92 @@ We will begin by creating a NotifyStep with its configuration and a doExecute me
...
-.. literalinclude:: ../../src/Acme/Bundle/NotifyConnectorBundle/Resources/config/batch_jobs.yml
+.. literalinclude:: ../../src/Acme/Bundle/NotifyConnectorBundle/Resources/config/job_profile.yml
    :language: yaml
    :linenos:
 
-You may have noticed that we use a custom class for the notification step and we define the handler as a step element.
+Translations
+------------
+
+Don't forget to add a translation to our new step:
```
Comment (rybus) [inline cookbook/import_export/create-custom-step.rst]: same here "don't forget". I am not very comfortable with this as "don't forget" means that readers had the occasion to know this befor,  which is not true. To me you should just write "Add a translation for our brand new step".

File: src/Acme/Bundle/NotifyConnectorBundle/Step/NotifyStep.php [hunk 13]
```diff
@@ -4,62 +4,27 @@
...
-    public function setHandler(CurlHandler $handler)
-    {
-        $this->handler= $handler;
-    }
-
-    // step items which are configurable with the job edit form
-    public function getConfigurableStepElements()
-    {
-        return array('handler' => $this->getHandler());
+        // we add custom details in summary
```
Comment (rybus) [inline src/Acme/Bundle/NotifyConnectorBundle/Step/NotifyStep.php]: in the summary

**Comment by rybus** [comment]
GTM @solivier !

**Pull request merged**